agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,551 @@
+            self.logger.info(
+                "Successfully wrote %s (%s.md5)",
+                host_tool_data_tb_name,
+                host_tool_data_tb_name,
+            )
+            # Tell the waiting "watcher" thread engaging in a "state" change
+            # that another PUT document has arrived.
+            with self._lock:
+                tm_tracker = self._tm_tracking[hostname]
+                assert tm_tracker["posted"] == "waiting", f"tm_tracker = {tm_tracker!r}"
+                tm_tracker["posted"] = "yes"
+                self._put_count += 1
+                self._cv.notify()
+
+
+def main(argv):
+    PROG = Path(argv[0]).name
+
+    logger = logging.getLogger(PROG)
+    fh = logging.FileHandler(f"{PROG}.log")
+    if os.environ.get("_PBENCH_UNIT_TESTS"):
+        fmtstr = "%(levelname)s %(name)s %(funcName)s -- %(message)s"
+    else:
+        fmtstr = (
+            "%(asctime)s %(levelname)s %(process)s %(thread)s"
+            " %(name)s %(funcName)s %(lineno)d -- %(message)s"
+        )
+    fhf = logging.Formatter(fmtstr)
+    fh.setFormatter(fhf)
+    fh.setLevel(logging.INFO)
+    logger.addHandler(fh)
+    logger.setLevel(logging.INFO)
+
+    try:
+        redis_host = argv[1]
+        redis_port = argv[2]
+        param_key = argv[3]
+    except IndexError as e:
+        logger.error("Invalid arguments: %s", e)
+        return 1
+
+    try:
+        redis_server = redis.Redis(host=redis_host, port=redis_port, db=0)
+    except Exception as e:
+        logger.error(
+            "Unable to connect to redis server, %s:%s: %s", redis_host, redis_port, e
+        )
+        return 2
+
+    try:
+        params_raw = redis_server.get(param_key)
+        if params_raw is None:
+            logger.error('Parameter key, "%s" does not exist.', param_key)
+            return 3
+        logger.info("params_key (%s): %r", param_key, params_raw)
+        params_str = params_raw.decode("utf-8")
+        # The expected parameters for this "data-sink" is what "channel" to
+        # subscribe to for the tool meister operational life-cycle.  The
+        # data-sink listens for the state transitions, start | stop | send |
+        # terminate, exiting when "terminate" is received, marking the state
+        # in which data is captured.
+        #
+        # E.g. params = '{ "channel": "run-chan",
+        #                  "benchmark_run_dir": "/loo/goo" }'
+        params = json.loads(params_str)
+        channel = params["channel"]
+        benchmark_run_dir = Path(params["benchmark_run_dir"]).resolve()
+    except Exception as ex:
+        logger.error("Unable to fetch and decode parameter key, %s: %s", param_key, ex)
+        return 4
+    else:
+        if not benchmark_run_dir.is_dir():
+            logger.error(
+                "Run directory argument, %s, must be a real" " directory.",
+                benchmark_run_dir,
+            )
+            return 5
+
+    pidfile_name = f"{PROG}.pid"
```

Comment:
Pidfiles should be stored in /var/run if we are following the LSB, otherwise the pid directory is going to be stored in the directory that the sink is going to run from.

Reply:
Using `/var/run` is definitely an option, but the decision to not use `/var/run` was based on the philosophy we want the footprint pbench places on a box to be minimal.  Having all the files and local state kept in one location allows it all to be easily collected and referred to later when debugging.